docs/design_proposals/debug-events.md

```diff
@@ -0,0 +1,187 @@
+
+With this information, the IDE should be able to:
+
+- Identify the corresponding artifact within the `skaffold.yaml`
+- Identify the corresponding artifact from a port-forward event by correlating the
+  remote port with the debug port from the previously-raised debug event
+- Establish new debug launch configurations with the forwarded port,
+  and the local source location, and any other necessary information
+- Tear down existing debug launch configurations when a container is terminated
+- (*In the future*, *if necessary*) Request port-forward of the debug port for the
+  namespace/pod/container/port on-demand
+ 
+### Example
+
+```json
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:38.916931Z",
+    "event": {
+      "deployEvent": {
+        "status": "In Progress"
+      }
+    },
+    "entry": "Deploy started"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:39.629826Z",
+    "event": {
+      "deployEvent": {
+        "status": "Complete"
+      }
+    },
+    "entry": "Deploy complete"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004396Z",
+    "event": {
+      "debugEvent": {
+        "status": "Started",
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "runtime": "nodejs",
+        "configuration": "{\"name\":\"gcr.io/artifact/name/from/skaffold.yaml\",\"ports\":{\"devtools\":9229},\"runtime\":\"nodejs\",\"workingDir\":\"/app\"}"
+      }
+    },
+    "entry": "Debuggable container started pod/npm:web (default)"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004551Z",
+    "event": {
+      "portEvent": {
+        "localPort": 3000,
+        "remotePort": 3000,
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "resourceType": "pod",
+        "resourceName": "npm"
+      }
+    },
+    "entry": "Forwarding container web to local port 3000"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004792Z",
+    "event": {
+      "portEvent": {
+        "localPort": 9229,
+        "remotePort": 9229,
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "portName": "devtools",
+        "resourceType": "pod",
+        "resourceName": "npm"
+      }
+    },
+    "entry": "Forwarding container web to local port 9229"
+  }
+}
+```
+
+### Open Issues/Questions
+
+**\<Why is a new Debug Event necessary?\>**
+
+One alternative explored was to piggy-back the debug information into the existing port-forward events.
+This design was eliminated as:
+ 
+  - There are some language runtimes, like .NET Core, that do not use debug ports and instead
+    require a `kubectl exec` to launch a process in the remote container. 
+  - Port-forwards may be dropped and re-established
+  
+There may also be a desire for the IDEs to selectively initiate port-forwards on an as-neeeded
```

Comment:
this is actually pretty huge. so we could just send a debug payload to skaffold, and request to open up a debug session on a port at any point during the dev session right? in theory, this would allow debugging during a normal `skaffold dev` loop wouldn't it? or is it required that we modify the container podspecs beforehand

Reply:
We'd still need to have modified the pod specs, though that could be done on-the-fly.